cores/esp8266/Schedule.cpp

```diff
@@ -152,28 +166,31 @@ void run_scheduled_recurrent_functions ()
         fence = true;
     }
 
-    recurrent_fn_t* prev = nullptr;
-    recurrent_fn_t* current = rFirst;
+    auto prev = &rFirst;
+    // prevent scheduling of new functions during this run
+    auto stop = rLast;
 
-    while (current)
+    bool done;
+    do
     {
-        if (current->callNow && !current->mFunc())
+        done = current == stop;
+        const bool wakeupToken = current->wakeupToken && current->wakeupToken->load();
+        const bool wakeup = current->wakeupTokenCmp != wakeupToken;
+        if (wakeup) current->wakeupTokenCmp = wakeupToken;
+        bool callNow = current->callNow;
+
+        if ((wakeup || callNow) && !current->mFunc())
         {
             // remove function from stack
             esp8266::InterruptLock lockAllInterruptsInThisScope;
 
             auto to_ditch = current;
 
-            if (prev)
-            {
-                current = current->mNext;
-                prev->mNext = current;
-            }
-            else
-            {
-                rFirst = rFirst->mNext;
-                current = rFirst;
-            }
+            // removing rLast
+            if (rLast == current) rLast = prev;
```

Comment:
No single-line if's please. Put the if body on its own line.

Reply:
Done.